### PR TITLE
[WIP] add protocol fees fixin

### PR DESCRIPTION
**Pull request by smarx**
## Description

This change adds `FixinProtocolFees`, code that collects protocol fees using per-pool `FeeCollector` contracts. Once per epoch, a function on each `FeeCollector` is called to transfer the held ether (first converted to WETH) to the staking contract. Delaying this transfer and associated accounting yields significant gas savings.

## Testing instructions

`yarn build && yarn test` in `contracts/zero-ex`.

The basic happy path with WETH is tested, but all other tests are missing. Future commits will add more robust testing, including negative testing for failure modes.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist:

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/zero-ex/contracts/src/external/FeeCollector.sol [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    /// @param poolId The pool ID this contract is collecting fees for.
+    function initialize(
+        IEtherTokenV06 weth,
+        IStaking staking,
+        bytes32 poolId
+    )
+        external
+        onlyAuthorized
+    {
+        weth.approve(address(staking), type(uint256).max);
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/external/FeeCollector.sol:48]: `type(uint256).max` :exploding_head:

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        } else {
+            // ETH
+            (bool success,) = address(feeCollector).call{value: amount}("");
+            require(success);
+        }
+    }
+
+    /// @dev Transfer fees for a given pool to the staking contract.
+    /// @param poolId Identifies the pool whose fees are being paid.
+    function _transferFeesForPool(
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:63]: Who will end up calling this?

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+pragma solidity ^0.6.5;
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-erc20/contracts/src/v06/IEtherTokenV06.sol";
+import "../external/FeeCollector.sol";
+
+/// @dev Helpers for collecting protocol fees.
+abstract contract FixinProtocolFees {
+    bytes32 immutable feeCollectorCodeHash;
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:27]: I guess this is the advantage over a library.

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol (hunk 2 above)
Comment (smarx) [reply contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:63]: Good question. I think the answer is that we need to expose it publicly and let _anyone_ call it. (0x Labs will likely be responsible for actually doing this at the end of every epoch. No one else has as much incentive to pay the gas fees to do it.)

So I could have perhaps made it an `external`, open function in the fixin itself. But I think it's better to expose it exactly once from a feature.

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol (hunk 3 above)
Comment (smarx) [reply contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:27]: Yes. I think this is the only advantage.
